src/PowerShell/PowerShellManagerPool.cs

```diff
@@ -15,37 +20,80 @@ namespace Microsoft.Azure.Functions.PowerShellWorker.PowerShell
     /// </summary>
     internal class PowerShellManagerPool
     {
-        private readonly ILogger _logger;
-        // Today we don't really support the in-proc concurrency. We just hold an instance of PowerShellManager in this field.
-        private PowerShellManager _psManager;
+        private readonly int _upperBound = 25;
+        private readonly MessagingStream _msgStream;
+        private readonly AsyncConcurrentPool<PowerShellManager> _pool;
+        private int _poolSize;
 
         /// <summary>
         /// Constructor of the pool.
         /// </summary>
-        internal PowerShellManagerPool(ILogger logger)
+        internal PowerShellManagerPool(MessagingStream msgStream)
         {
-            _logger = logger;
+            string upperBound = Environment.GetEnvironmentVariable("InProcConcurrencyUpperBound");
+            if (string.IsNullOrEmpty(upperBound) || !int.TryParse(upperBound, out _upperBound))
+            {
+                _upperBound = 1;
+            }
+
+            _msgStream = msgStream;
+            _pool = new AsyncConcurrentPool<PowerShellManager>();
         }
 
         /// <summary>
         /// Initialize the pool and populate it with PowerShellManager instances.
-        /// When it's time to really implement this pool, we probably should instantiate PowerShellManager instances in a lazy way.
-        /// Maybe start from size 1 and increase the number of workers as needed.
+        /// We instantiate PowerShellManager instances in a lazy way, starting from size 1 and increase the number of workers as needed.
         /// </summary>
-        internal void Initialize()
+        internal void Initialize(string requestId)
         {
-            _psManager = new PowerShellManager(_logger);
+            var logger = new RpcLogger(_msgStream);
+
+            try
+            {
+                logger.SetContext(requestId, invocationId: null);
+                _pool.Add(new PowerShellManager(logger));
+                _poolSize = 1;
+            }
+            finally
+            {
+                logger.ResetContext();

```

Comment:
Why is the ResetContext needed here if we instantiate a new RpcLogger every time?